pkg/parser/parser.go

```diff
@@ -0,0 +1,291 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package parser
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"strings"
+
+	"github.com/spf13/afero"
+	corev1 "k8s.io/api/core/v1"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes"
+
+	apiextensionsv1alpha1 "github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	pkgmetav1alpha1 "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
+)
+
+// Parser is a package parser.
+type Parser interface {
+	Parse(context.Context, ...BackendOption) (*Package, error)
+}
+
+// parser is a Parser implementation with pluggable backends.
+type parser struct {
+	backend Backend
+}
+
+// New returns a new parser.
+func New(backend Backend) Parser {
+	return &parser{
+		backend: backend,
+	}
+}
+
+// BackendOption modifies the parser backend. Backends may accept options at
+// creation time, but must accept them at initialization.
+type BackendOption func(Backend)
+
+// Backend provides a source for the parser.
+type Backend interface {
+	Init(context.Context, ...BackendOption) (io.ReadCloser, error)
+}
+
+// Package includes fields that all meaningful parsers should support.
+type Package struct {
+	provider                     *pkgmetav1alpha1.Provider
+	configuration                *pkgmetav1alpha1.Configuration
+	customResourceDefinitions    map[string]*apiextensions.CustomResourceDefinition
+	compositeResourceDefinitions map[string]*apiextensionsv1alpha1.CompositeResourceDefinition
+	compositions                 map[string]*apiextensionsv1alpha1.Composition
+}
+
+// NewPackage returns a new Package with maps initialized.
+func NewPackage() *Package {
+	return &Package{
+		customResourceDefinitions:    map[string]*apiextensions.CustomResourceDefinition{},
+		compositeResourceDefinitions: map[string]*apiextensionsv1alpha1.CompositeResourceDefinition{},
+		compositions:                 map[string]*apiextensionsv1alpha1.Composition{},
+	}
+}
+
+// GetProvider gets the package provider manifest.
+func (p *Package) GetProvider() *pkgmetav1alpha1.Provider {
+	return p.provider
+}
+
+// GetConfiguration gets the package configuration manifest.
+func (p *Package) GetConfiguration() *pkgmetav1alpha1.Configuration {
+	return p.configuration
+}
```

Comment:
Is there something this package could to do help the caller determine whether the Package contains a Provider or a Configuration? I was thinking something like a `func (p *Package) IsProvider() bool` convenience method.

Reply:
I had thought about that, but felt that `IsProvider` might be too restrictive if we want to potentially have more than 2 package types in the future. The idea I had been playing around with was to have a `PackageType()` method that returned the GVK of the `Provider`, `Configuration`, or future package type.

Reply:
Yeah, a `PackageType` method that returned a GVK seems more future proof.